cmd/identity/main.go

```diff
@@ -21,13 +26,132 @@ var (
+		Args:  cobra.ExactArgs(1),
+		RunE:  cmdCSR,
+	}
+
+	config struct {
+		BaseDir        string `default:"$CONFDIR/identity" help:"Directory containing service subdirectories"`
+		Difficulty     uint64 `default:"15" help:"minimum difficulty for identity generation"`
+		Timeout        string `default:"5m" help:"timeout for CA generation; golang duration string (0 no timeout)"`
+		Concurrency    uint   `default:"4" help:"number of concurrent workers for certificate authority generation"`
+		ParentCertPath string `help:"path to the parent authority's certificate chain"`
+		ParentKeyPath  string `help:"path to the parent authority's private key"`
+		Signer         certificates.CertClientConfig
+	}
+
 	defaultConfDir = fpath.ApplicationDir("storj", "identity")
 )
 
+func init() {
+	rootCmd.AddCommand(newServiceCmd)
+	rootCmd.AddCommand(csrCmd)
+
+	cfgstruct.Bind(newServiceCmd.Flags(), &config, cfgstruct.ConfDir(defaultConfDir))
+	cfgstruct.Bind(csrCmd.Flags(), &config, cfgstruct.ConfDir(defaultConfDir))
+}
+
 func main() {
 	process.Exec(rootCmd)
 }
 
+func cmdNewService(cmd *cobra.Command, args []string) error {
+	serviceErrs := new(errs.Group)
+	if len(args) > 0 {
+		for _, service := range args {
+			serviceDir := filepath.Join(config.BaseDir, service)
+			caCertPath := filepath.Join(serviceDir, "ca.cert")
+			caKeyPath := filepath.Join(serviceDir, "ca.key")
+			identCertPath := filepath.Join(serviceDir, "identity.cert")
+			identKeyPath := filepath.Join(serviceDir, "identity.key")
+
+			ca, err := identity.CASetupConfig{
+				CertPath: caCertPath,
+				KeyPath:  caKeyPath,
+			}.Create(process.Ctx(cmd))
+			if err != nil {
+				serviceErrs.Add(err)
+			}
+
+			_, err = identity.SetupConfig{
+				CertPath: identCertPath,
+				KeyPath:  identKeyPath,
+			}.Create(ca)
+			if err != nil {
+				serviceErrs.Add(err)
+			}
+		}
+	}
+
+	return serviceErrs.Err()
+}
+
+func cmdCSR(cmd *cobra.Command, args []string) error {
+	ctx := process.Ctx(cmd)
+
+	serviceDir := filepath.Join(config.BaseDir, args[0])
+	caCertPath := filepath.Join(serviceDir, "ca.cert")
+	caKeyPath := filepath.Join(serviceDir, "ca.key")
+	caConfig := identity.FullCAConfig{
+		CertPath: caCertPath,
+		KeyPath:  caKeyPath,
+	}
+	identCertPath := filepath.Join(serviceDir, "identity.cert")
+	identKeyPath := filepath.Join(serviceDir, "identity.key")
+	identConfig := identity.Config{
+		CertPath: identCertPath,
+		KeyPath:  identKeyPath,
+	}
+
+	ca, err := caConfig.Load()
+	if err != nil {
+		return err
+	}
+	ident, err := identConfig.Load()
+	if err != nil {
+		return err
+	}
+
+	signedChainBytes, err := config.Signer.Sign(ctx, ident)
+	if err != nil {
+		return errs.New("error occurred while signing certificate: %s\n(identity files were still generated and saved, if you try again existing files will be loaded)", err)
+	}
+
+	signedChain, err := identity.ParseCertChain(signedChainBytes)
+	if err != nil {
+		return nil
+	}
+
+	ca.Cert = signedChain[0]
+	ca.RestChain = signedChain[1:]
```

Comment:
it shouldn't have only 1 item because it should contain the cert that was sent to be signed and the cert that signed it; however, even if it did i don't think this would be an error.